pkg/workspace/controllers/workspace_controller.go

```diff
@@ -666,45 +666,73 @@ func (c *WorkspaceReconciler) applyInference(ctx context.Context, wObj *kaitov1b
-					deployment := existingObj.(*appsv1.Deployment)
-					if deployment.Annotations[kaitov1beta1.WorkspaceRevisionAnnotation] != revisionStr {
-						// TODO: respect updates to other fields
-						var volumes []corev1.Volume
-						var volumeMounts []corev1.VolumeMount
-						shmVolume, shmVolumeMount := utils.ConfigSHMVolume()
-						if shmVolume.Name != "" {
-							volumes = append(volumes, shmVolume)
-						}
-						if shmVolumeMount.Name != "" {
-							volumeMounts = append(volumeMounts, shmVolumeMount)
-						}
-
-						if len(wObj.Inference.Adapters) > 0 {
-							adapterVolume, adapterVolumeMount := utils.ConfigAdapterVolume()
-							volumes = append(volumes, adapterVolume)
-							volumeMounts = append(volumeMounts, adapterVolumeMount)
-						}
-
-						pullerContainers, pullerEnvVars, pullerVolumes := manifests.GeneratePullerContainers(wObj, volumeMounts)
-						volumes = append(volumes, pullerVolumes...)
-
-						spec := &deployment.Spec.Template.Spec
-						spec.Containers[0].Env = pullerEnvVars
-						spec.Containers[0].VolumeMounts = volumeMounts
-						spec.InitContainers = pullerContainers
-						spec.Volumes = volumes
-
-						deployment.Annotations[kaitov1beta1.WorkspaceRevisionAnnotation] = revisionStr
-
-						if err := c.Update(ctx, deployment); err != nil {
-							return
-						}
+				annotations := existingObj.GetAnnotations()
+				if annotations == nil {
+					annotations = make(map[string]string)
+				}
+
+				currentRevisionStr, ok := annotations[kaitov1beta1.WorkspaceRevisionAnnotation]
+				if !ok || currentRevisionStr != revisionStr {
+					// TODO: respect updates to other fields
+					var volumes []corev1.Volume
+					var volumeMounts []corev1.VolumeMount
+					shmVolume, shmVolumeMount := utils.ConfigSHMVolume()
+					if shmVolume.Name != "" {
+						volumes = append(volumes, shmVolume)
+					}
+					if shmVolumeMount.Name != "" {
+						volumeMounts = append(volumeMounts, shmVolumeMount)
+					}
+
+					if len(wObj.Inference.Adapters) > 0 {
+						adapterVolume, adapterVolumeMount := utils.ConfigAdapterVolume()
+						volumes = append(volumes, adapterVolume)
+						volumeMounts = append(volumeMounts, adapterVolumeMount)
+					}
+
+					pullerContainers, pullerEnvVars, pullerVolumes := manifests.GeneratePullerContainers(wObj, volumeMounts)
+					volumes = append(volumes, pullerVolumes...)
+
+					var spec *corev1.PodSpec
+					switch existingObj := existingObj.(type) {
+					case *appsv1.StatefulSet:
+						spec = &existingObj.Spec.Template.Spec
+					case *appsv1.Deployment:
+						spec = &existingObj.Spec.Template.Spec
+					}
+
+					// Perform upsert instead of replacing to avoid losing existing
+					// env vars, volume mounts, and volumes that are not related to the puller.
+					spec.Containers[0].Env = utils.Upsert(
+						spec.Containers[0].Env,
+						pullerEnvVars,
+						func(env corev1.EnvVar) string { return env.Name },
+					)
+					spec.Containers[0].VolumeMounts = utils.Upsert(
```

Comment:
If users delete the adapter, will you still keep the original adapter?

Reply:
No.. it's a bug. Let me fix that